knative-operator/pkg/controller/knativeservingobsolete/knativeservingobsolete_controller.go

```diff
@@ -87,31 +94,113 @@ func (r *ReconcileKnativeServingObsolete) Reconcile(request reconcile.Request) (
 			return reconcile.Result{}, err
 		}
 	}
-	// Create the latest CR from the current (previous) CR
-	latest = &servingv1alpha1.KnativeServing{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      current.Name,
-			Namespace: current.Namespace,
-		},
-	}
-	latest.Spec.Config = current.Spec.Config
-	if err := common.Mutate(latest, r.client); err != nil {
+	// Orphan all the children by removing their OwnerRefs
+	if err := r.orphanObsoleteResources(current); err != nil {
 		return reconcile.Result{}, err
 	}
 	// Avoid a certs config conflict in the k-s controller
 	if err := r.removeOldCertsConfig(current.Namespace); err != nil {
 		return reconcile.Result{}, err
 	}
-	// Orphan the kids to avoid webhook race condition
-	if err := r.client.Delete(context.TODO(), current, client.PropagationPolicy(metav1.DeletePropagationOrphan)); err != nil {
+
+	new, err := r.reconcileNewResource(current)
+	if err != nil {
 		return reconcile.Result{}, err
 	}
-	if err := r.client.Create(context.TODO(), latest); err != nil {
-		return reconcile.Result{}, err
+
+	wantStatus := obsolete.KnativeServingStatus{
+		Version:    new.Status.Version,
+		Conditions: deepCopyConditions(new.Status.Conditions),
+	}
+	if !equality.Semantic.DeepEqual(current.Status, wantStatus) {
+		current.Status = wantStatus
+		if err := r.client.Status().Update(context.TODO(), current); err != nil {
+			return reconcile.Result{}, err
+		}
 	}
+
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileKnativeServingObsolete) reconcileNewResource(old *obsolete.KnativeServing) (*servingv1alpha1.KnativeServing, error) {
+	new := &servingv1alpha1.KnativeServing{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: old.Namespace, Name: old.Name}, new)
+	if errors.IsNotFound(err) {
+		new := &servingv1alpha1.KnativeServing{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      old.Name,
+				Namespace: old.Namespace,
+			},
+			Spec: servingv1alpha1.KnativeServingSpec{
+				Config: old.Spec.Config,
+			},
+		}
+		if err := common.Mutate(new, r.client); err != nil {
+			return nil, err
+		}
+		if err := r.client.Create(context.TODO(), new); err != nil {
+			return nil, err
+		}
+		return new, nil
+	} else if err != nil {
+		return nil, err
+	} else {
+		want := new.DeepCopy()
+		want.Spec.Config = old.Spec.Config
+		if err := common.Mutate(want, r.client); err != nil {
```

Comment:
I have issues with this. I'm concerned that manual edits to the new config will be overwritten by manual edits to the old config. To me, that violates the principle of least surprise. I think the conservative approach is to only update new from old with fields not already configured.

Reply:
I hear you, though I think that is a somewhat dangerous assumption to begin with. My plan was to document that we're doing a "mirror migration" and if you want to start using the new CRD you **must** delete the old one first to avoid exactly those kinda situations.

Does that work for you as well?

Reply:
How about we add a validation webhook for the old which prevents any spec updates with an error referencing the new CR? That would support old CR creation and prevent updates with an error message encouraging migration.

Reply:
I think that'd potentially also breaks support of the old CR, though I'm not sure if any of the systems we have actually update the old CR. I could imagine the Console to do that 🤔  but I don't know for sure.

Reply:
Well, if you must, how do you feel about replacing the "want/Mutate" logic with this instead?
https://gist.github.com/4b867fb15da60adb08e68fe24bf5df49

Reply:
BTW, that logic ^^ might obviate your need to add the istio config to Mutate, which is probably harmless, but I'd prefer to have conditioned on *something*, I think.

Reply:
That's super elegant, taken!

I think I'd like us to keep the serviceMesh config in there. I agree it somewhat breaks an abstraction, but we don't really need an abstraction there anyway. IMO, config shall be done completely via the CR. In any case.